src/lib/CoreInstance.cc

```diff
@@ -0,0 +1,229 @@
+  // Currently, only a flat memory interface can be used for the instruction
+  // memory
+  if (type != simeng::MemInterfaceType::Flat) {
+    std::cerr
+        << "Incompatible non-flat instruction memory interface type requested"
+        << std::endl;
+    exit(1);
+  }
+
+  // Create a L1I cache instance based on type supplied
+  if (type == simeng::MemInterfaceType::Flat) {
+    instructionMemory_ = std::make_shared<simeng::FlatMemoryInterface>(
+        processMemory_.get(), processMemorySize_);
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    instructionMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["L1-Cache"]["Access-Latency"].as<uint16_t>());
+  }
+
+  return instructionMemory_;
+}
+
+void CoreInstance::setL1InstructionMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1I cache instance to use
+  instructionMemory_ = memRef;
+  return;
+}
+
+std::shared_ptr<simeng::MemoryInterface> CoreInstance::createL1DataMemory(
+    const simeng::MemInterfaceType type) {
+  // Currently, if the core in use is emulation or in-order, only a flat data
+  // memory interface can be used
+  if (mode_ == SimulationMode::Emulation &&
+      type != simeng::MemInterfaceType::Flat) {
+    std::cerr << "Incompatible non-flat data memory interface type requested "
+                 "with emulation core"
+              << std::endl;
+    exit(1);
+  }
+  if (mode_ == SimulationMode::InOrderPipelined &&
+      type != simeng::MemInterfaceType::Flat) {
+    std::cerr << "Incompatible non-flat data memory interface type requested "
+                 "with in-order core"
+              << std::endl;
+    exit(1);
+  }
+
+  // Create a L1D cache instance based on type supplied
+  if (type == simeng::MemInterfaceType::Flat) {
+    dataMemory_ = std::make_shared<simeng::FlatMemoryInterface>(
+        processMemory_.get(), processMemorySize_);
+  } else if (type == simeng::MemInterfaceType::Fixed) {
+    dataMemory_ = std::make_shared<simeng::FixedLatencyMemoryInterface>(
+        processMemory_.get(), processMemorySize_,
+        config_["L1-Cache"]["Access-Latency"].as<uint16_t>());
+  }
+
+  return dataMemory_;
+}
+
+void CoreInstance::setL1DataMemory(
+    std::shared_ptr<simeng::MemoryInterface> memRef) {
+  // Set the L1D cache instance to use
+  dataMemory_ = memRef;
+  return;
+}
+
+std::shared_ptr<simeng::Core> CoreInstance::createCore() {
+  // Ensure all appropriate creation functions have been called
+  if (instructionMemory_ == nullptr) {
+    std::cerr << "Instruction memory not instantiated or set" << std::endl;
+    exit(1);
+  } else if (dataMemory_ == nullptr) {
+    std::cerr << "Data memory not instantiated or set" << std::endl;
+    exit(1);
+  } else if (process_ == nullptr) {
+    std::cerr << "Process not instantiated or set" << std::endl;
+    exit(1);
+  }
```

Comment:
I'm trying to reason about why this is needed over just using the object's constructor.

Sureley we could pass the  process's path string to the constructor and memory types can be inferred from the config file rather than having to make so many function calls? And as for default program, this could be defined in a header file to access as needed (if no string is passed to constructor).

Reply:
After previous comment (https://github.com/UoB-HPC/SimEng/pull/247#discussion_r970890787) I understand that this decision comes from SST compatability. Setting as resolved, no clarification needed. #

Reply:
So are you suggesting we have extra config file params for defining the memory hierarchy which can be used to configure the memory objects within CoreInstance? Not a bad plan and would work well with future in-house cache solutions.

For the process image, we could do something equivalent to that. May be better to keep two separate constructors to avoid complications between an incorrect executable path of `""` and a way to denote 'use hex span'.

Reply:
Resolving due to the significant changes made in the latest commits. Please re-review.